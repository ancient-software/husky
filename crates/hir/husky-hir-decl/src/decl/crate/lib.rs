use super::*;

#[salsa::interned]
pub struct LibCrateHirDecl {}
