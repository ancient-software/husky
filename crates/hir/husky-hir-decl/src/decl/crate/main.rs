use super::*;

#[salsa::interned]
pub struct MainCrateHirDecl {}
