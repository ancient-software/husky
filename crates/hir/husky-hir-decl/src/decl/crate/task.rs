use super::*;

#[salsa::interned]
pub struct TaskCrateHirDecl {}
