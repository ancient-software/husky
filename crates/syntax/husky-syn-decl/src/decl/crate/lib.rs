#[path = "lib/narrative.rs"]
pub mod narrative;

use self::narrative::LibCrateSynDeclNarrative;
use super::*;
use husky_crate_decl_ast::{CrateDeclAst, CrateDeclAstIdxRange};
use parsec::{PunctuatedSmallList, TryParseOptionFromStream};

#[salsa::tracked]
pub struct LibCrateSynNodeDecl {
    #[id]
    pub path: CratePath,
    #[return_ref]
    pub items: Vec<SynNodeDeclResult<LibCrateSynDeclItem>>,
    pub syn_expr_region: SynExprRegion,
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum LibCrateSynDeclItem {
    Narrative {
        narrate_token: NarrateRegionalToken,
        narrative: LibCrateSynDeclNarrative,
    },
}

impl<'db> CrateDeclParser<'db> {
    pub(super) fn parse_lib_crate_syn_node_decl(mut self) -> LibCrateSynNodeDecl {
        let db = self.db();
        let crate_path = self.crate_path();
        use husky_print_utils::p;
        use salsa::DebugWithDb;
        p!(crate_path.debug(db));
        let crate_decl_tokra_region_data = self.crate_decl_tokra_region_data();
        let mut items = vec![];
        for ast_idx in crate_decl_tokra_region_data.root_body() {
            items.push(match crate_decl_tokra_region_data[ast_idx] {
                CrateDeclAst::Err => todo!(),
                CrateDeclAst::BasicStmtOrBranch {
                    regional_token_verse_idx,
                    body,
                } => self.parse_basic_stmt(regional_token_verse_idx, body),
                CrateDeclAst::IfElseStmts {
                    if_branch,
                    elif_branches,
                    else_branch,
                } => todo!(),
                CrateDeclAst::MatchStmt {
                    regional_token_verse_idx,
                    pattern_stmt,
                    case_branches,
                } => todo!(),
            })
        }
        let syn_expr_region = self.finish();
        LibCrateSynNodeDecl::new(db, crate_path, items, syn_expr_region)
    }

    fn parse_basic_stmt(
        &mut self,
        regional_token_verse_idx: RegionalTokenVerseIdx,
        body: Option<CrateDeclAstIdxRange>,
    ) -> SynNodeDeclResult<LibCrateSynDeclItem> {
        let mut parser = self.token_verse_expr_parser(regional_token_verse_idx);
        if let Some(basic_stmt_kw) = parser.try_parse_option::<BasicStmtLeadRegionalToken>()? {
            match basic_stmt_kw {
                BasicStmtLeadRegionalToken::Let(_) => todo!(),
                BasicStmtLeadRegionalToken::Return(_) => todo!(),
                BasicStmtLeadRegionalToken::Require(_) => todo!(),
                BasicStmtLeadRegionalToken::Assert(_) => todo!(),
                BasicStmtLeadRegionalToken::Break(_) => todo!(),
                BasicStmtLeadRegionalToken::For(_) => todo!(),
                BasicStmtLeadRegionalToken::ForExt(_) => todo!(),
                BasicStmtLeadRegionalToken::While(_) => todo!(),
                BasicStmtLeadRegionalToken::Do(_) => todo!(),
                BasicStmtLeadRegionalToken::Narrate(narrate_token) => {
                    self::narrative::parse_narrative(parser, narrate_token)
                }
            }
        } else {
            todo!()
        }
    }
}

#[salsa::tracked]
pub struct LibCrateSynDecl {
    #[id]
    pub path: CratePath,
    #[return_ref]
    pub items: Vec<LibCrateSynDeclItem>,
    pub syn_expr_region: SynExprRegion,
}

impl LibCrateSynDecl {
    pub(super) fn from_node(
        path: CratePath,
        syn_node_decl: LibCrateSynNodeDecl,
        db: &::salsa::Db,
    ) -> SynDeclResult<Self> {
        let items = syn_node_decl
            .items(db)
            .iter()
            .map(|result| -> SynDeclResult<LibCrateSynDeclItem> {
                result.as_ref().map(Clone::clone).map_err(Into::into)
            })
            .collect::<SynDeclResult<Vec<LibCrateSynDeclItem>>>()?;
        let syn_expr_region = syn_node_decl.syn_expr_region(db);
        Ok(Self::new(db, path, items, syn_expr_region))
    }
}
