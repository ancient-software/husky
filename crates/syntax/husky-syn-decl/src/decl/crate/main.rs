use super::*;

#[salsa::tracked]
pub struct MainCrateSynNodeDecl {
    #[id]
    pub path: CratePath,
    pub syn_expr_region: SynExprRegion,
}

impl<'db> CrateDeclParser<'db> {
    pub(super) fn parse_main_crate_syn_node_decl(self) -> MainCrateSynNodeDecl {
        todo!()
    }
}

#[salsa::tracked]
pub struct MainCrateSynDecl {
    #[id]
    pub path: CratePath,
    pub syn_expr_region: SynExprRegion,
}

impl MainCrateSynDecl {
    pub(super) fn from_node(
        path: CratePath,
        syn_node_decl: MainCrateSynNodeDecl,
        db: &::salsa::Db,
    ) -> SynDeclResult<Self> {
        let syn_expr_region = syn_node_decl.syn_expr_region(db);
        Ok(Self::new(db, path, syn_expr_region))
    }
}
