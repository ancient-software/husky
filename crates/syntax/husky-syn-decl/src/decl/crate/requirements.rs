use super::*;

#[salsa::tracked]
pub struct RequirementsCrateSynNodeDecl {
    #[id]
    pub path: CratePath,
    pub syn_expr_region: SynExprRegion,
}

impl<'db> CrateDeclParser<'db> {
    pub(super) fn parse_requirements_crate_syn_node_decl(&self) -> RequirementsCrateSynNodeDecl {
        todo!()
    }
}
