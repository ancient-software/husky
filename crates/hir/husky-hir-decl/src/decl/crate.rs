#[path = "crate/lib.rs"]
pub mod lib;
#[path = "crate/main.rs"]
pub mod main;
#[path = "crate/task.rs"]
pub mod task;

use self::lib::*;
use self::main::*;
use self::task::*;
use super::*;

pub enum CrateHirDecl {
    Lib(LibCrateHirDecl),
    Main(MainCrateHirDecl),
}
