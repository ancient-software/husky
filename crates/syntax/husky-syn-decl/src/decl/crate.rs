#[path = "crate/lib.rs"]
pub mod lib;
#[path = "crate/main.rs"]
pub mod main;
#[path = "crate/requirements.rs"]
pub mod requirements;
#[path = "crate/task.rs"]
pub mod task;

use self::lib::*;
use self::main::*;
use self::requirements::*;
use self::task::*;
use super::*;
use husky_vfs::{CrateKind, CratePath};

#[salsa::derive_debug_with_db]
#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash)]
pub enum CrateSynNodeDecl {
    Lib(LibCrateSynNodeDecl),
    Main(MainCrateSynNodeDecl),
    Requirements(RequirementsCrateSynNodeDecl),
    Task(TaskCrateSynNodeDecl),
}

impl HasSynNodeDecl for CratePath {
    type NodeDecl = CrateSynNodeDecl;

    fn syn_node_decl<'a>(self, db: &'a salsa::Db) -> Self::NodeDecl {
        crate_syn_node_decl(db, self)
    }
}

#[salsa::tracked]
fn crate_syn_node_decl(db: &::salsa::Db, crate_path: CratePath) -> CrateSynNodeDecl {
    let parser = CrateDeclParser::new(db, crate_path);
    parser.parse_crate_syn_node_decl()
}

impl<'db> CrateDeclParser<'db> {
    fn parse_crate_syn_node_decl(&self) -> CrateSynNodeDecl {
        let db = self.db();
        match self.crate_path().kind(db) {
            CrateKind::Lib => self.parse_lib_crate_syn_node_decl(),
            CrateKind::Main => self.parse_lib_crate_syn_node_decl(),
            CrateKind::Requirements => self.parse_lib_crate_syn_node_decl(),
            CrateKind::Task => self.parse_lib_crate_syn_node_decl(),
            CrateKind::Bin(_) => todo!(),
            CrateKind::IntegratedTest(_) => todo!(),
            CrateKind::Example => todo!(),
        }
    }
}

#[test]
fn crate_syn_node_decl_works() {
    DB::ast_expect_test_debug_with_db(
        crate_syn_node_decl,
        &AstTestConfig::new(
            "crate_syn_node_decl",
            FileExtensionConfig::Markdown,
            TestDomainsConfig::SYNTAX,
        ),
    )
}
