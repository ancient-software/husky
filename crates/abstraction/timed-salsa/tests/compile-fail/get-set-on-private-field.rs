#[timed_salsa::jar(db = Db)]
pub struct Jar(a::MyInput);

mod a {
    use crate::Jar;

    #[timed_salsa::input(jar = Jar)]
    pub struct MyInput {
        field: u32,
    }
}

pub trait Db: timed_salsa::DbWithJar<Jar> {}

#[timed_salsa::db(Jar)]
#[derive(Default)]
struct Database {
    storage: timed_salsa::Storage<Self>,
}

impl timed_salsa::Database for Database {}

impl Db for Database {}

fn main() {
    let mut db = Database::default();
    let input = a::MyInput::new(&mut db, 22);

    input.field(&db);
    input.set_field(&mut db).to(23);
}
