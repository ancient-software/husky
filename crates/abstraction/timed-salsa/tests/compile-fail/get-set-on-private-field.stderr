error[E0624]: associated function `field` is private
  --> tests/compile-fail/get-set-on-private-field.rs:29:11
   |
7  |     #[timed_salsa::input(jar = Jar)]
   |     -------------------------- private associated function defined here
...
29 |     input.field(&db);
   |           ^^^^^ private associated function

error[E0624]: associated function `set_field` is private
  --> tests/compile-fail/get-set-on-private-field.rs:30:11
   |
7  |     #[timed_salsa::input(jar = Jar)]
   |     -------------------------- private associated function defined here
...
30 |     input.set_field(&mut db).to(23);
   |           ^^^^^^^^^ private associated function
